### PR TITLE
feat: Implement system protocol

### PR DESCRIPTION
**Pull request by baxen**
See the example in tests/systems.rs

I've also put some documentatino in that previews an idea around macros, but they are not yet implemented

### CODE REVIEW

File: crates/goose/src/providers/types/tool.rs [hunk 1]
```diff
@@ -1,140 +1,123 @@
...
+pub struct ToolParameter {
+    /// Description of what the parameter does
+    pub description: String,
+    /// The type of the parameter (e.g., "string", "number", "boolean")
+    pub parameter_type: String,
+    /// Whether this parameter is required
+    pub required: bool,
+}
+
+impl ToolParameter {
```
Comment (salman1993) [inline crates/goose/src/providers/types/tool.rs]: can this support arrays and objects? we can have a tool params like this:
```
{
  "parameters": {
    "type": "object",
    "properties": {
      "action": {
        "type": "string",
        "description": "The action to be performed, such as displaying FAQs"
      },
      "faq_entries": {
        "type": "array",
        "description": "A list of FAQ entries, each with a question and an answer",
        "items": {
          "type": "object",
          "properties": {
            "question": {
              "type": "string",
              "description": "The question text for the FAQ entry"
            },
            "answer": {
              "type": "string",
              "description": "The answer text for the FAQ entry"
            }
          },
          "required": ["question", "answer"]
        }
      }
    },
    "required": ["action", "faq_entries"]
  }
}
```

File: docs/systems.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+- A description
+- A set of parameters
+- An implementation that executes the tool's functionality
+
+Tools are defined using a custom attribute macro:
+
+```rust
+#[tool(
+    name = "echo",
+    description = "Echoes back the input"
```
Comment (salman1993) [inline docs/systems.md]: is this missing `parameters` or that's intentional? i was thinking it'd be this:
```
#[tool(
    name = "echo",
    description = "Echoes back the input",
    parameters = {
        "message": ("string", "required", "The message to echo back")
    }
)]
```

File: docs/systems.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+### Example Implementation
+
+Here's a complete example of a simple system:
+
+```rust
+use goose_macros::tool;
+
+struct FileSystem {
+    registry: ToolRegistry,
```
Comment (salman1993) [inline docs/systems.md:76]: is the ToolRegistry meant to be shared across all the systems (1 registry for all tools) or one per system (1 registry for every system's tools)?

File: crates/goose/src/providers/types/tool.rs (hunk 1 above)
Comment (baxen) [reply crates/goose/src/providers/types/tool.rs]: yeah good point, let me think about this - might have to give up some of the type safety here and go back to how you had it with just Value

File: docs/systems.md (hunk 2 above)
Comment (baxen) [reply docs/systems.md]: no you're right would have to add that here too

File: docs/systems.md (hunk 3 above)
Comment (baxen) [reply docs/systems.md:76]: in this example it'd be a private (maybe even entirely hidden) member of the system as a way to implement the macro driven approach. E.g. it'd be a generalization of this from the test system

```
    async fn call_tool(&self, tool_use: ToolUse) -> ToolResult<Value> {
        match tool_use.name.as_str() {
            "echo" => self.echo(tool_use.parameters).await,
            _ => ToolResult::Err(ToolError::ToolNotFound(tool_use.name)),
        }
    }
```

**Comment by michaelneale** [comment]
missing a dep? 

```
Caused by:
  failed to load manifest for dependency `goose-macros`

Caused by:
  failed to read `/Users/micn/Documents/code/goose/crates/goose-macros/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
```

**Comment by baxen** [comment]
> missing a dep?
> 
> ```
> Caused by:
>   failed to load manifest for dependency `goose-macros`
> 
> Caused by:
>   failed to read `/Users/micn/Documents/code/goose/crates/goose-macros/Cargo.toml`
> 
> Caused by:
>   No such file or directory (os error 2)
> ```

Removing for now - i had a POC but will wait on it, since we don't actually plan to build more than a handful of rust native systems. this can be a followup to 1.0 release likely

**Pull request merged**